### PR TITLE
Add ProcessInstanceModificationProcessor

### LINKED ISSUE #9639: Process a modification command
## Description

We need the basic infrastructure to process a modification command. This is the initial step before implementing the logic to apply the modification for the process instance. 

* create a new processor for the new modification command
* write a `MODIFIED` event when receiving a `MODIFY` command
* send a `MODIFIED` response to the client 

Blocked by #9638

### PR DESCRIPTION
**Pull request by remcowesterhoud**
## Description

Adds the processor to enable us to start processing PI modification commands.

## Related issues

closes #9639

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  @Override
+  public boolean onCommand(
+      final TypedRecord<ProcessInstanceModificationRecord> command,
+      final CommandControl<ProcessInstanceModificationRecord> controller) {
+    final ProcessInstanceModificationRecord record = command.getValue();
+
+    stateWriter.appendFollowUpEvent(
+        keyGenerator.nextKey(), ProcessInstanceModificationIntent.MODIFIED, record);
+
+    controller.accept(ProcessInstanceModificationIntent.MODIFIED, record);
```
Comment (remcowesterhoud) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java:32]: ❓ @reviewer I am not sure if this is correct, so please correct me if it's not. My goal is to write the `MODIFIED` event and after send a `MODIFIED` response back to the client, like mentioned in the issue:

- write a MODIFIED event when receiving a MODIFY command
- send a MODIFIED response to the client

Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java:39]: Yes, `controller.accept` takes care of this for you. It writes that event and (depending on the returned boolean of your `onCommand` method implementation) sends it also as a response.

The reason for this is that you've implemented the `ProcessInstanceModificationProcessor` as a `CommandProcessor`. When this is registered in `ProcessEventProcessors` it uses an overloaded method of `TypedRecordProcessors.onCommand`, that wraps your `CommandProcessor` implementation with the `CommandProcessorImpl`. This is a wrapper that tries to generify what a command processor does (i.e. write an event, and send a response). 

Alternatively, `ProcessInstanceModificationProcessor` could implement `TypedRecordProcessor` directly. In that case the registration happen over another of the overloaded methods, avoiding the wrapper. (Note the wrapper is an implementation of `TypedRecordProcessor` itself).

The thing about this wrapper is that it adds indirection in an attempt to clarify what a command processor should do. However, it does so in a rather unclear way, by using this command control and depending on the return value of the implemented method and such. It also restricts the number of records that a command processor can write (at least through the command control mechanism).

With previous changes, we've removed some of the usages of this wrapper, but this effort hasn't been completed. IMO we should avoid it for new processors.

I recommend that you make `ProcessInstanceModificationProcessor` implement `TypedRecordProcessor` directly.

**Review by korthout — APPROVED** [review]
Thanks @remcowesterhoud 🚀 

🔧 This works correctly, but I'd recommend taking the time to implement it differently. Please have a look at my comment

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java (hunk 1 above)
Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java:39]: So my current solution would then write the event twice? I'll change it like you said, but that's still good to know 😄

**Comment by remcowesterhoud** [comment]
Thanks @korthout I have changed this, please have another look 🙂

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java (hunk 1 above)
Comment (korthout) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java]: Oh wow, you're right. I missed that completely 😆 Goes to show how confusing this wrapper is

File: engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+  @Override
+  public void processRecord(final TypedRecord<ProcessInstanceModificationRecord> command) {
+    final ProcessInstanceModificationRecord record = command.getValue();
+
+    stateWriter.appendFollowUpEvent(
+        keyGenerator.nextKey(), ProcessInstanceModificationIntent.MODIFIED, record);
+
+    responseWriter.writeEventOnCommand(
+        keyGenerator.nextKey(), ProcessInstanceModificationIntent.MODIFIED, record, command);
```
Comment (korthout) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java:41]: ❌ Please make sure to reuse the generated key.

Comment (remcowesterhoud) [reply engine/src/main/java/io/camunda/zeebe/engine/processing/processinstance/ProcessInstanceModificationProcessor.java]: My bad 🙈

**Comment by remcowesterhoud** [comment]
bors merge

**Comment by remcowesterhoud** [comment]
bors retry

**Pull request merged**